### ISSUE TITLE
tensorflow serving not working

### ISSUE THREAD

**Issue by surajbharadwaj** [zzh8829/yolov3-tf2#28]
it gives an error when making a request to server. Says binary and op are different. Does this model work when server is loaded?

**Comment by Thoorium** [comment]
Hi,

The current release of Tensorflow Serving is not up to date with the alpha of Tensorflow 2, meaning that some OPs are incompatible or don't exist between the versions.

For now, you can use the nightly docker release of Tensorflow Serving https://www.tensorflow.org/tfx/serving/docker to run TF2 models. Replace the docker image in the documentation "tensorflow/serving" with "tensorflow/serving:nightly".

**Comment by surajbharadwaj** [comment]
Ohh really? I will try that. And do you have a client script for it? should it be converted into base64 and sent as json POST? If you do have a script can you please share it?
Thank You

**Comment by Thoorium** [comment]
I'm still working on the client portion of the process pipeline. As far as I know, the image needs to be sent as a normalized float array.

**Comment by surajbharadwaj** [comment]
https://github.com/Kammerlo/keras-yolo3-serving/blob/master/serving/http_client.py
https://github.com/Kammerlo/keras-yolo3-serving/blob/master/serving/export_tf_serving.py
This guy is changing his input layer to accept base64 before exporting to tf serving format file.
Do you need to something similar to this and would it work?

**Comment by Thoorium** [comment]
Yes you can add a signature to process the image on the server and accept a base64 image instead. In Tensorflow 2, it would be something like this:

```python
@tf.function(input_signature=[tf.TensorSpec((None), tf.string)])
def decode_image(input_image):
    image = tf.dtypes.cast(tf.io.decode_image(input_image[0]), dtype=tf.float32)
    t = tf.convert_to_tensor(image, np.float32)
    t = t / 255.0
    return t

signatures = {"serving_default": decode_image}

# Export the model with the signature
tf.saved_model.save(model, "saved_model", signatures)
```

You can take a look at `export_tfserving.py` and modify the signature there to fit your use case.

**Comment by 17702513221** [comment]
when i use docker run -p 8501:8501 --mount type=bind,source=/home/xs/yolov3-tf2/serving/yolov3,target=/models/yolov3 -e MODEL_NAME=yolov3 -t tensorflow/serving:nightly
result:2019-07-03 01:47:04.253741: I tensorflow_serving/model_servers/server.cc:350] Running gRPC ModelServer at 0.0.0.0:8500 ...
[warn] getaddrinfo: address family for nodename not supported
2019-07-03 01:47:04.254404: I tensorflow_serving/model_servers/server.cc:370] Exporting HTTP/REST API at:localhost:8501 ...
[evhttp_server.cc : 239] RAW: Entering the event loop ...
Could you give me a client to test,thankyou

**Comment by Thoorium** [comment]
Something I quickly put together. I strongly recommend using a base64 signature instead.
```python
import numpy as np
import tensorflow as tf
import requests
import sys
np.set_printoptions(threshold=sys.maxsize) # Allow numpy to fully print out the array as a string, skipping the ellipsis

img = tf.image.decode_image(open("PATH/TO/YOUR/IMAGE/HERE", 'rb').read(), channels=3)
img = tf.expand_dims(img, 0)
img = tf.image.resize(img, (416, 416))
img = img / 255
arr = np.array2string(img.numpy(), separator=',')
url = 'http://localhost:8501/v1/models/yolov3:predict'
data = '{"instances": ' + arr + '}'
response = requests.post(url, data=data)
```
Your can also use tools like Postman for quick tests with Restful api

**Comment by 17702513221** [comment]
> Something I quickly put together. I strongly recommend using a base64 signature instead.
> 
> ```python
> import numpy as np
> import tensorflow as tf
> import requests
> import sys
> np.set_printoptions(threshold=sys.maxsize) # Allow numpy to fully print out the array as a string, skipping the ellipsis
> 
> img = tf.image.decode_image(open("PATH/TO/YOUR/IMAGE/HERE", 'rb').read(), channels=3)
> img = tf.expand_dims(img, 0)
> img = tf.image.resize(img, (416, 416))
> img = img / 255
> arr = np.array2string(img.numpy(), separator=',')
> url = 'http://localhost:8501/v1/models/yolov3:predict'
> data = '{"instances": ' + arr + '}'
> response = requests.post(url, data=data)
> ```
> 
> Your can also use tools like Postman for quick tests with Restful api

The results is 
2019-07-05 15:54:28.197693: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-07-05 15:54:28.230900: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2496000000 Hz
2019-07-05 15:54:28.231285: I tensorflow/compiler/xla/service/service.cc:162] XLA service 0x34621b0 executing computations on platform Host. Devices:
2019-07-05 15:54:28.231320: I tensorflow/compiler/xla/service/service.cc:169]   StreamExecutor device (0): <undefined>, <undefined>

**Comment by hugovaz** [comment]
Has anyone been using base64 as input? I've been trying but I'm unable to get the input_signature for base64 pasted here to work with the model. I don't have much experience with this part and I'm struggling  a bit.

**Comment by zzh8829** [comment]
I believe TF Serving should work now, closing the issue

**Comment by bishoymoussa** [comment]
> > Something I quickly put together. I strongly recommend using a base64 signature instead.
> > ```python
> > import numpy as np
> > import tensorflow as tf
> > import requests
> > import sys
> > np.set_printoptions(threshold=sys.maxsize) # Allow numpy to fully print out the array as a string, skipping the ellipsis
> > 
> > img = tf.image.decode_image(open("PATH/TO/YOUR/IMAGE/HERE", 'rb').read(), channels=3)
> > img = tf.expand_dims(img, 0)
> > img = tf.image.resize(img, (416, 416))
> > img = img / 255
> > arr = np.array2string(img.numpy(), separator=',')
> > url = 'http://localhost:8501/v1/models/yolov3:predict'
> > data = '{"instances": ' + arr + '}'
> > response = requests.post(url, data=data)
> > ```
> > 
> > 
> > Your can also use tools like Postman for quick tests with Restful api
> 
> The results is
> 2019-07-05 15:54:28.197693: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
> 2019-07-05 15:54:28.230900: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2496000000 Hz
> 2019-07-05 15:54:28.231285: I tensorflow/compiler/xla/service/service.cc:162] XLA service 0x34621b0 executing computations on platform Host. Devices:
> 2019-07-05 15:54:28.231320: I tensorflow/compiler/xla/service/service.cc:169] StreamExecutor device (0): ,

I have the same issue

**Comment by taheritajar** [comment]
> Something I quickly put together. I strongly recommend using a base64 signature instead.
> 
> ```python
> import numpy as np
> import tensorflow as tf
> import requests
> import sys
> np.set_printoptions(threshold=sys.maxsize) # Allow numpy to fully print out the array as a string, skipping the ellipsis
> 
> img = tf.image.decode_image(open("PATH/TO/YOUR/IMAGE/HERE", 'rb').read(), channels=3)
> img = tf.expand_dims(img, 0)
> img = tf.image.resize(img, (416, 416))
> img = img / 255
> arr = np.array2string(img.numpy(), separator=',')
> url = 'http://localhost:8501/v1/models/yolov3:predict'
> data = '{"instances": ' + arr + '}'
> response = requests.post(url, data=data)
> ```
> 
> Your can also use tools like Postman for quick tests with Restful api

How can I use Postman as a Client of serving?
